include/seastar/parquet/column_chunk_reader.hh

```diff
@@ -0,0 +1,396 @@
+            : std::optional<uint32_t>{}}
+        {};
+    // Read a batch of n (rep, def, value) triplets. The last batch may be smaller than n.
+    // Return the number of triplets read. Note that null values are not read into the output array.
+    // Example output: def == [1, 1, 0, 1, 0], rep = [0, 0, 0, 0, 0], val = ["a", "b", "d"].
+    template<typename LevelT>
+    seastar::future<size_t> read_batch(size_t n, LevelT def[], LevelT rep[], output_type val[]);
+};
+
+template<format::Type::type T>
+template<typename LevelT>
+seastar::future<size_t>
+column_chunk_reader<T>::read_batch_internal(size_t n, LevelT def[], LevelT rep[], output_type val[]) {
+    if (_eof) {
+        return seastar::make_ready_future<size_t>(0);
+    }
+    if (!_initialized) {
+        return load_next_page().then([this, n, def, rep, val] {
+            return read_batch_internal(n, def, rep, val);
+        });
+    }
+    size_t def_levels_read = _def_decoder.read_batch(n, def);
+    size_t rep_levels_read = _rep_decoder.read_batch(n, rep);
+    if (def_levels_read != rep_levels_read) {
+        throw parquet_exception::corrupted_file(seastar::format(
+                "Number of definition levels {} does not equal the number of repetition levels {} in batch",
+                def_levels_read, rep_levels_read));
+    }
+    if (def_levels_read == 0) {
+        _initialized = false;
+        return read_batch_internal(n, def, rep, val);
+    }
+    for (size_t i = 0; i < def_levels_read; ++i) {
+        if (def[i] < 0 || def[i] > static_cast<LevelT>(_schema_node.def_level)) {
+            throw parquet_exception::corrupted_file(seastar::format(
+                    "Definition level ({}) out of range (0 to {})", def[i], _schema_node.def_level));
+        }
+        if (rep[i] < 0 || rep[i] > static_cast<LevelT>(_schema_node.rep_level)) {
+            throw parquet_exception::corrupted_file(seastar::format(
+                    "Repetition level ({}) out of range (0 to {})", rep[i], _schema_node.rep_level));
+        }
+    }
+    size_t values_to_read = 0;
+    for (size_t i = 0; i < def_levels_read; ++i) {
+        if (def[i] == static_cast<LevelT>(_schema_node.def_level)) {
+            ++values_to_read;
+        }
+    }
+    size_t values_read = _val_decoder.read_batch(values_to_read, val);
+    if (values_read != values_to_read) {
+        throw parquet_exception::corrupted_file(seastar::format(
+                "Number of values in batch {} is less than indicated by def levels {}", values_read, values_to_read));
+    }
+    return seastar::make_ready_future<size_t>(def_levels_read);
+}
+
+template<format::Type::type T>
+template<typename LevelT>
+seastar::future<size_t>
+inline column_chunk_reader<T>::read_batch(size_t n, LevelT def[], LevelT rep[], output_type val[]) {
+    return seastar::futurize_apply([this, n, def, rep, val] {
+        return read_batch_internal(n, def, rep, val);
+    }).handle_exception([this] (std::exception_ptr eptr) {
```

Comment:
If you made sure that read_batch_internal is noexcept (by returning failed futures instead), you could also get rid of the `futurize_apply` wrapper.